jiant/utils/python/datastructures.py

```diff
@@ -1,29 +1,73 @@
 import math
 import itertools
-from typing import Mapping, Any, Sequence, Iterable, Iterator, Union
+from typing import Mapping, Any, Sequence, Iterable, Iterator, Union, Tuple, Dict
 
 
 def take_one(ls: Union[Sequence, Mapping]) -> Any:
+    """Extract item from a collection containing a just one item.
+
+    Args:
+        ls (Union[Sequence, Mapping]): collection containing a single item.
+
+    Returns:
+        Single item from collection.
+
+    """
     if not len(ls) == 1:
         raise IndexError(f"has more than one element ({len(ls)})")
     return next(iter(ls))
 
 
 def chain_idx_get(container: Union[Sequence, Mapping], key_list: Sequence, default: Any) -> Any:
+    """Retrieve entry at a path from a arbitrarily nested collection, return default if not found.
+
+    Args:
+        container (Union[Sequence, Mapping]): collection from which to try to retrieve element.
+        key_list (Sequence): list of index and/or keys specifying the path to the requested element.
+        default (Any): default value to return if no value exists at the specified path.
+
+    Returns:
+        Entry found at the specified path (or default value if no entry is found at that path).
+
+    """
     try:
         return chain_idx(container, key_list)
     except (KeyError, IndexError, TypeError):
         return default
 
 
 def chain_idx(container: Union[Sequence, Mapping], key_list: Sequence) -> Any:
+    """Retrieve entry at a path from a arbitrarily nested collection.
+
+    Args:
+        container (Union[Sequence, Mapping]): collection from which to try to retrieve element.
+        key_list (Sequence): list of index and/or keys specifying the path to the requested element.
+
+    Returns:
+        Entry found at the specified path.
+
+    """
     curr = container
     for key in key_list:
         curr = curr[key]
     return curr
 
 
 def group_by(ls: Sequence, key_func) -> dict:
+    """Apply a function to every element of a sequence.
```

Comment:
This description is a little misleading. The purpose of `group_by` is to group elements by the result of the function applied to each element. This example might be clearer:

```python
group_by([1, 2, 3, 4, 5], lambda x: x%2==0)
# Output: {False: [1, 3, 5], True: [2, 4]}
```

Reply:
Thanks, copied this explanation and your example directly into the code (in 3a8c3df).